### PR TITLE
add some new ops, fix some operators and add batch operations to certain operators.

### PR DESCRIPTION
**Pull request by leejet**
I tested it in the [sd.cpp](https://github.com/leejet/stable-diffusion.cpp) project, and everything works well.

### CODE REVIEW

**Comment by FSSRepo** [comment]
@leejet I believe that before merging this, the respective Metal kernels should be created to maintain compatibility, although it doesn't really help much since sd.cpp doesn't benefit from the Metal backend as matrix operations are very slow.

**Comment by leejet** [comment]
OK,  I will find time to add support for Metal backend.

**Comment by leejet** [comment]
@FSSRepo I have implemented Metal backend support for ggml_arange and ggml_timestep_embedding, and tested it with the test-arrange and test-timestep-embedding test cases to ensure it works correctly. Could you find some time to review and merge this PR?

**Comment by FSSRepo** [comment]
Ok, I will test this PR with latest sd.cpp

**Comment by slaren** [comment]
- Does `arange` affect the performance significantly? Are there any downsides to constructing the tensor in a CPU buffer and copying it to the backend as an input?
- Does `timestep_embedding` have any use outside of SD? And again, would it hurt performance significantly if it was implemented on the CPU and copied to the backend as an input?

**Comment by leejet** [comment]
These two operations won't significantly improve performance, but in stable video diffusion, it's necessary to perform these operations multiple times within the UNet network. Additionally, the parameters may change with each execution. If I don't execute these operations when computing the graph, I would have to remember the tensors produced by each operation and their corresponding parameters. Then, during the ggml_gallocr_alloc_graph, I'd have to copy the data, making the process more complex. Implementing this through ggml operators would make the code more elegant.

File: src/ggml-cuda.cu
Comment (slaren) [inline src/ggml-cuda.cu:1083]: This is using double precision functions, which can be very slow on some devices. Using `expf`, `cosf` and `sinf` instead should improve performance.

File: src/ggml-cuda.cu
Comment (slaren) [inline src/ggml-cuda.cu:10575]: This is not necessary, with ggml-backend all the tensors are always allocated on the device.

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:5924]:
```suggestion
```

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:5938]:
```suggestion
    int actual_dim = dim;
```

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:5947]: Use `ggml_set_op_params` or `ggml_set_op_params_i32`.

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:5920]: Use `ggml_set_op_params`.

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:13624]: Use `memcpy` instead.

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:13668]: Use `ggml_get_op_params_i32`.

File: src/ggml.c
Comment (slaren) [inline src/ggml.c:13682]: Also might want to use single precision math functions here for performance.

File: tests/test-timestep_embedding.cpp
Comment (slaren) [inline tests/test-timestep_embedding.cpp:1]: It seems that this test can never fail.

**Review by slaren — COMMENTED** [review]
It would be good to have tests for these operators in `test-backend-ops`.

File: include/ggml/ggml.h
Comment (ggerganov) [inline include/ggml/ggml.h:1691]:
```suggestion
            struct ggml_context * ctx,
            float                 start,
            float                 stop,
            float                 step);

```

File: src/ggml-cuda.cu
Comment (ggerganov) [inline src/ggml-cuda.cu:10559]:
```suggestion
static void ggml_cuda_arange(const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst) {
    ggml_tensor_extra_gpu * dst_extra = (ggml_tensor_extra_gpu *)  dst->extra;

    const bool dst_on_device = dst->backend == GGML_BACKEND_TYPE_GPU;
```

File: src/ggml-metal.m
Comment (ggerganov) [inline src/ggml-metal.m:2322]:
```suggestion
                        [encoder setComputePipelineState:pipeline];
                        [encoder setBuffer:id_dst  offset:offs_dst    atIndex:0];
                        [encoder setBytes:&ne0   length:sizeof(ne0)   atIndex:1];
                        [encoder setBytes:&start length:sizeof(start) atIndex:2];
                        [encoder setBytes:&step  length:sizeof(step)  atIndex:3];
```

File: src/ggml-metal.m
Comment (ggerganov) [inline src/ggml-metal.m:2343]:
```suggestion
                        [encoder setComputePipelineState:pipeline];
                        [encoder setBuffer:id_src0 offset:offs_src0 atIndex:0];
                        [encoder setBuffer:id_dst  offset:offs_dst  atIndex:1];
                        [encoder setBytes:&nb1   length:sizeof(nb1) atIndex:2];
                        [encoder setBytes:&dim   length:sizeof(dim) atIndex:3];
                        [encoder setBytes:&max_period length:sizeof(max_period) atIndex:4];
```

File: src/ggml-metal.metal
Comment (ggerganov) [inline src/ggml-metal.metal:1971]:
```suggestion
    device float * dst_ptr = (device float *) dst;
```

File: src/ggml.c
Comment (ggerganov) [inline src/ggml.c:10322]: What was the reason to change these normalization to be inside the loop?

**Review by ggerganov — APPROVED** [review]
After adding tests to `test-backend-ops` we can merge

File: src/ggml.c
Comment (leejet) [reply src/ggml.c:10322]: Because previously, it calculated the sum of (v * v) and then divided by (ne00 * ne01 * step), I was concerned that in certain situations, it might lead to overflow or precision issues. Therefore, I changed it to calculate the sum of (v * v)/(ne00 * ne01 * step) to obtain the variance.

File: src/ggml-cuda.cu
Comment (leejet) [reply src/ggml-cuda.cu]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c:5934]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: src/ggml.c
Comment (leejet) [reply src/ggml.c]: Changed in the latest commit.

File: tests/test-timestep_embedding.cpp
Comment (leejet) [reply tests/test-timestep_embedding.cpp:1]: Changed in the latest commit.

File: src/ggml-cuda.cu
Comment (leejet) [reply src/ggml-cuda.cu:10575]: This part was copied from ggml_cuda_op_flatten, and I'm not sure whether to make changes in this PR or create a separate PR to modify both places?

**Comment by leejet** [comment]
I have modified the code based on the suggestions from the review and added tests to test-backend-ops. There are two uncertain points still awaiting feedback from the maintainer.

File: src/ggml.c
Comment (ggerganov) [reply src/ggml.c]: It shouldn't cause problems since we accumulate into `double` (i.e. `ggml_float`). I've introduced a separate accumulator per row and restored the normalization to be at the end in order to avoid extra division in the loop

**Review by ggerganov — APPROVED** [review]

**Review by slaren — APPROVED** [review]

**Pull request merged**
